clients/src/main/java/org/apache/kafka/clients/producer/ProducerConfig.java

```diff
@@ -514,12 +512,11 @@ public ProducerConfig(Map<String, Object> props) {
     }
 
     boolean idempotenceEnabled() {
-        boolean userConfiguredIdempotence = this.originals().containsKey(ENABLE_IDEMPOTENCE_CONFIG);
         boolean userConfiguredTransactions = this.originals().containsKey(TRANSACTIONAL_ID_CONFIG);
-        boolean idempotenceEnabled = userConfiguredIdempotence && this.getBoolean(ENABLE_IDEMPOTENCE_CONFIG);
-
-        if (!idempotenceEnabled && userConfiguredIdempotence && userConfiguredTransactions)
+        boolean idempotenceEnabled = this.getBoolean(ENABLE_IDEMPOTENCE_CONFIG);
+        if (!idempotenceEnabled && userConfiguredTransactions)
             throw new ConfigException("Cannot set a " + ProducerConfig.TRANSACTIONAL_ID_CONFIG + " without also enabling idempotence.");
+
         return userConfiguredTransactions || idempotenceEnabled;
```

Comment:
nit: I think the check for `userConfiguredTransactions` is redundant now.

Reply:
You are right! After we default the `enable.idempotence` to true, there's no way to have `idempotenceEnabled==false` but `userConfiguredTransactions==true`. Remove it.